Improve metadata detection

R-Palazzo opened the pull request:
Resolve #1515
- Compared to the issue, I just added one condition for `Object` detection:
```
if len(data) <= 10:
    sdtype = 'categorical'
```
I did it because otherwise a lot of test we're failing due to columns detected as `id` rather than `categorical` now. I also feel that this follow a similar statement that we have for numbers which is:
`If there are <=5 rows, we can't properly detect this. Sdtype is numerical.
`
Let me know if this works @npatki 

- One other thing, I noticed there is a lot of `staticmethod` in `SingleTableMetadata`. Following the [google styleguide](https://google.github.io/styleguide/pyguide.html), do we want to move them outside of the class in this PR? @amontanez24, I did it first but then I realised it was confusing which changes are proper to this issue or just cleaning so I removed it. Let me know what we prefer.

amontanez24 commented:
> Resolve #1515
> 
> * Compared to the issue, I just added one condition for `Object` detection:
> 
> ```
> if len(data) <= 10:
>     sdtype = 'categorical'
> ```
> 
> I did it because otherwise a lot of test we're failing due to columns detected as `id` rather than `categorical` now. I also feel that this follow a similar statement that we have for numbers which is: `If there are <=5 rows, we can't properly detect this. Sdtype is numerical. ` Let me know if this works @npatki
> 
> * One other thing, I noticed there is a lot of `staticmethod` in `SingleTableMetadata`. Following the [google styleguide](https://google.github.io/styleguide/pyguide.html), do we want to move them outside of the class in this PR? @amontanez24, I did it first but then I realised it was confusing which changes are proper to this issue or just cleaning so I removed it. Let me know what we prefer.

You can just leave the staticmethods. We can also alter our personal style guide to tolerate them

pvk-developer commented on sdv/metadata/single_table.py:
```diff
@@ -245,11 +244,92 @@ def to_dict(self):
+        sdtype = None
+        try:
+            with warnings.catch_warnings():
+                warnings.simplefilter('ignore', category=UserWarning)
```
What warnings are being shown here and why we ignore them? 🤔

R-Palazzo replied:
It was because we didn't give any `datetime_format` and we ignored it because the user couldn't do anything about it. But now I removed it after the changes of [4aa4a4a](https://github.com/sdv-dev/SDV/pull/1529/commits/4aa4a4ac2c96acb7a1fa5929cc5826512e128fc8)

pvk-developer commented on sdv/metadata/single_table.py:
```diff
@@ -245,11 +244,92 @@ def to_dict(self):
+        try:
+            with warnings.catch_warnings():
+                warnings.simplefilter('ignore', category=UserWarning)
+                pd.to_datetime(data)
```
There is one more requirement towards this: All `datetimes` should be in the same `format` . My suggestion to this is:
1. Detect the datetime format.
2. Cast to `pd.to_datetime(data, datetime_fromat=detected_format, errors='raise')`
3. If error is raised then we ignore the datetimes.

pvk-developer commented on tests/integration/multi_table/test_hma.py:
```diff
@@ -604,6 +604,21 @@ def test_use_own_data_using_hma(tmp_path):
         sdtype='id',
         regex_format='HID_[0-9]{3,4}'
     )
+    metadata.update_column(
+        table_name='hotels',
+        column_name='city',
+        sdtype='categorical',
+    )
+    metadata.update_column(
+        table_name='hotels',
+        column_name='state',
+        sdtype='categorical',
+    )
+    metadata.update_column(
+        table_name='hotels',
+        column_name='classification',
+        sdtype='categorical',
+    )
```
Are we updating this because now they are no longer `categories` but `ids` ?

R-Palazzo replied:
They were detected as `unknown` but it was mostly because the demo dataset here is small and so the condition `Elif the cardinality is less than round(len / 5), sdtype is categorical` was not reached. I discussed it with Neha and we decided that it was fine like this.

pvk-developer reviewed (changes requested):
I think we have to change something on the datetime but other than that looks good ;)

R-Palazzo commented:
Thanks for your review @pvk-developer. I addressed the comments in [4aa4a4a](https://github.com/sdv-dev/SDV/pull/1529/commits/4aa4a4ac2c96acb7a1fa5929cc5826512e128fc8).

amontanez24 commented on tests/integration/single_table/test_copulas.py:
```diff
@@ -361,3 +362,35 @@ def test_numerical_columns_gets_pii():
+    # Setup
+    data = pd.DataFrame({
+        'category_col': [
+            1, 2, 1, 2, 1, 2, np.nan, 1, 1, np.nan, 2, 2, None, 2,
+            1, 1, None, 1, 2, 2
```
Just so you know, pandas converts the Nones here to np.nan which is why the dtype stays float. They aren't really necessary because of that. If ypou want the `None` values to stay you have to manually set the dtype to object

R-Palazzo replied:
Yes I put `np.nan` and `None` now as a sanity check that everything works for both.

amontanez24 replied:
No I mean that even though you put both in here, pandas converts the Nones to np.nan behind the scenes so None is not actually being tested. To test it you have to set the dtype to object

R-Palazzo commented on sdv/metadata/single_table.py:
```diff
@@ -245,11 +244,90 @@ def to_dict(self):
+        """
+        sdtype = None
+        try:
+            datetime_format = get_datetime_format(data)
+            pd.to_datetime(data, format=datetime_format, errors='raise')
```
I thought that we need to ensure here that all the data is convertible with the same datetime format. Should we subsample? I didn't look at the computational time.

amontanez24 replied:
The goal is to check that everything matches, but since it can possibly be slow for very large datasets, we can probably pick a max number of rows and say if at least that many match it's probably fine

amontanez24 commented on sdv/metadata/single_table.py:
```diff
@@ -247,11 +246,96 @@ def to_dict(self):
+                    f"Unsupported data type for column '{field}' (kind: {dtype})."
+                    "The valid data types are: 'object', 'int', 'float', 'datetime', 'bool'."
+                )
+            column_dict = {'sdtype': deepcopy(sdtype)}
```
minor: do we need to deepcopy here? Shouldn't sdtype be a string?

amontanez24 commented on sdv/metadata/single_table.py:
```diff
@@ -247,11 +246,96 @@ def to_dict(self):
+                column_dict['pii'] = True
+            elif sdtype == 'datetime':
+                datetime_format = get_datetime_format(clean_data)
+                column_dict['datetime_format'] = datetime_format
```
I think we said we'd only set the datetime format in the case where we parsed it from the text. In other words
- If the dtype is datetime
    - make the `sdtype` datetime and the `datetime_format` None
- If the dtype is object
    - if it can be converted to a datetime set the `sdtype` to datetime and set the `datetime_format`

You can maybe have the method that parses the datetime return the format if it finds it or something like that

R-Palazzo replied:
alright I see thanks, done in [58e4fcf](https://github.com/sdv-dev/SDV/pull/1529/commits/58e4fcf49eff5120b51b460bd92b594371086c43)

frances-h reviewed (approved):
Looks good!
